numpy/_core/numeric.py

```diff
@@ -2508,6 +2507,60 @@ def array_equiv(a1, a2):
     return bool(asarray(a1 == a2).all())
 
 
+def _astype_dispatcher(x, dtype, /, *, copy=None):
+    return (x, dtype)
+
+
+@array_function_dispatch(_astype_dispatcher)
+def astype(x, dtype, /, *, copy = True):
+    """
+    Copies an array to a specified data type.
+
+    This function is an Array API compatible alternative to
+    `numpy.ndarray.astype`.
+
+    Parameters
+    ----------
+    x : array_like
+        Input array to cast.
+    dtype : dtype
+        Data type of the result.
+    copy : bool, optional
+        Specifies whether to copy an array when the specified dtype matches
+        the data type of the input array ``x``. If ``True``, a newly allocated
+        array must always be returned. If ``False`` and the specified dtype
+        matches the data type of the input array, the input array must be
+        returned; otherwise, a newly allocated array must be returned.
+        Defaults to ``True``.
+
+    Returns
+    -------
+    out : ndarray
+        An array having the specified data type.
+
+    See Also
+    --------
+    ndarray.astype
+
+    Examples
+    --------
+    >>> arr = np.array([1, 2, 3]); arr
+    array([1, 2, 3])
+    >>> np.astype(arr, np.float64)
+    array([1., 2., 3.])
+
+    Non-copy case:
+
+    >>> arr = np.array([1, 2, 3])
+    >>> arr_noncpy = np.astype(arr, arr.dtype, copy=False)
+    >>> np.shares_memory(arr, arr_noncpy)
+    True
+
+    """
+    x = asarray(x)
```

Comment:
This isn't really right:
* You would have to use `copy` in the `asarray` call.
* The `copy=copy` in `astype()` doesn't have the same semantics.

I am actually not sure if `astype` should *coerce* from to an array at all?  If this does a forced cast, then `astype` *becomes* `asarray()`.

Reply:
I also wonder if any coercion should be allowed here and I would agree if it shouldn't.

Reply:
I think I agree about no coercion here. Instead, it seems to me like this should verify that the input is already an array.

That would leave the question of how exactly to do that. The dispatcher for `__array_function__` was already added, so I think it should work to check with `isinstance(x, np.ndarray)` (allows subclasses, but not duck arrays - those should be using `__array_function__`).

And, if we do that, what's the right exception type? I'd think `TypeError`?

Reply:
I second `TypeError`. Its documentation says:
```
Raised when an operation or function is applied to an object of inappropriate type.